libimage/runtime.go

```diff
@@ -414,61 +431,34 @@ func (r *Runtime) lookupImageInLocalStorage(name, candidate string, options *Loo
 	return image, nil
 }
 
-// lookupImageInDigestsAndRepoTags attempts to match name against any image in
-// the local containers storage.  If name is digested, it will be compared
-// against image digests.  Otherwise, it will be looked up in the repo tags.
-func (r *Runtime) lookupImageInDigestsAndRepoTags(name string, options *LookupImageOptions) (*Image, string, error) {
-	// Until now, we've tried very hard to find an image but now it is time
-	// for limbo.  If the image includes a digest that we couldn't detect
-	// verbatim in the storage, we must have a look at all digests of all
-	// images.  Those may change over time (e.g., via manifest lists).
-	// Both Podman and Buildah want us to do that dance.
-	allImages, err := r.ListImages(context.Background(), nil, nil)
-	if err != nil {
-		return nil, "", err
-	}
-
-	ref, err := reference.Parse(name) // Warning! This is not ParseNormalizedNamed
-	if err != nil {
-		return nil, "", err
-	}
-	named, isNamed := ref.(reference.Named)
-	if !isNamed {
+// lookupImageInRepoTags attempts to match named against any image in the local
+// containers storage by comparing the name to all known repo tags.
+func (r *Runtime) lookupImageInRepoTags(name string, namedName reference.Named, options *LookupImageOptions) (*Image, string, error) {
+	if namedName == nil {
 		return nil, "", fmt.Errorf("%s: %w", name, storage.ErrImageUnknown)
 	}
 
-	digested, isDigested := named.(reference.Digested)
-	if isDigested {
-		logrus.Debug("Looking for image with matching recorded digests")
-		digest := digested.Digest()
-		for _, image := range allImages {
-			for _, d := range image.Digests() {
-				if d != digest {
-					continue
-				}
-				// Also make sure that the matching image fits all criteria (e.g., manifest list).
-				if _, err := r.lookupImageInLocalStorage(name, image.ID(), options); err != nil {
-					return nil, "", err
-				}
-				return image, name, nil
-
-			}
-		}
+	if _, isDigested := namedName.(reference.Digested); isDigested {
```

Comment:
(This is a fairly strong restriction.

For tagged inputs, `inRepoTags` accepts any suffix match, i.e. given an input `fedora`, `example.com/fedora` and `example.com/ns/fedora` both match (without `example.com…` being in the short name search path at all).

I think it would be more consistent for this function to apply the same repo rules for digest references as well, instead of going to the full extreme of requiring a precise repo match.

That’s what is failing in https://github.com/containers/podman/pull/18888 .

OTOH I certainly won’t claim that any of `lookupImageInRepoTags` is a good idea, so if you are fine with breaking those inputs, \*shrug\*.

Arguably, if we go from accepting _any_ repo to accepting the same repos as for tagged inputs, that’s _still_ breaking compatibility, so I really have no reason to argue that we _need_ to accept the same repos as for tagged inputs; it’s a middle ground between breaking everything and not breaking anything that is not better than either of the two.)

Reply:
I see your line of argument.  The repo-dance should apply equally to digested and tagged references.

Reply:
I mean… I really don’t want to strongly argue for the tag/digest logic equivalence here. It would be cleaner if we started from scratch, but given current state, _changing_ to get that outcome is not obviously preferable to any of the other options.

If the consensus to this point has been to do a strict repo check, I’m fine with going with that, instead of restarting the discussion from scratch again.